### PR TITLE
cadence-workflow 0.18.2 (new formula)

### PR DESCRIPTION
**Pull request by longquanzheng**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

```
(cadence)$brew audit --strict cadence-workflow
qlong@/usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula:
(cadence)$echo $?
0
```
-----

### CODE REVIEW

File: Formula/cadence-workflow.rb [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  desc "Cadence Workflow is a distributed, scalable, durable, and highly available orchestration engine to execute asynchronous long-running business logic in a scalable and resilient way"
+  homepage "https://cadenceworkflow.io/"
+  url "https://github.com/uber/cadence.git",
+      tag:      "v0.18.1",
+      revision: "b6dfe37ebfec7742b1470be7129f07518e8bec98"
+  license "MIT"
+
+  bottle do
+  end
+
```
Comment (carlocab) [inline Formula/cadence-workflow.rb:11]:
```suggestion
```
You don't need this. @BrewTestBot will add this upon merging.

File: Formula/cadence-workflow.rb [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  depends_on "go" => :build
+
+  conflicts_with "cadence", because: "both install an `cadence` executable"
+
+  def install
+    ENV["PATH"] = "#{ENV["PATH"]}:#{ENV["GOPATH"]}/bin"
```
Comment (carlocab) [inline Formula/cadence-workflow.rb:17]: Why is this needed?

File: Formula/cadence-workflow.rb [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    system "make", "cadence", "cadence-server", "cadence-canary", "cadence-sql-tool", "cadence-cassandra-tool"
+    bin.install "cadence"
+    bin.install "cadence-server"
+    bin.install "cadence-canary"
+    bin.install "cadence-sql-tool"
+    bin.install "cadence-cassandra-tool"
```
Comment (carlocab) [inline Formula/cadence-workflow.rb:23]: Is there no `make install`? We should use that.

File: Formula/cadence-workflow.rb [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+    system "cp", "-r", "config", "#{etc}/cadence"
+    system "cp", "-r", "schema", "#{etc}/cadence"
```
Comment (carlocab) [inline Formula/cadence-workflow.rb:26]:
```suggestion
    (etc/"cadence").install "config", "schema"
```

File: Formula/cadence-workflow.rb [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+  end
+
+  test do
+    assert_match "cadence:",
+      shell_output("#{bin}/cadence version")
```
Comment (carlocab) [inline Formula/cadence-workflow.rb:31]: Can you propose a more substantial test than checking the version or calling for help? These tests rarely ever fail even for broken installations.

>We want tests that don’t require any user input and test the basic functionality of the application. For example `foo build-foo input.foo` is a good test and (despite their widespread use) `foo --version` and `foo --help` are bad tests. 

Ref. https://docs.brew.sh/Formula-Cookbook#add-a-test-to-the-formula

File: Formula/cadence-workflow.rb [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+  bottle do
+  end
+
+  depends_on "go" => :build
+
+  conflicts_with "cadence", because: "both install an `cadence` executable"
+
+  def install
+    ENV["PATH"] = "#{ENV["PATH"]}:#{ENV["GOPATH"]}/bin"
+    system "make", "cadence", "cadence-server", "cadence-canary", "cadence-sql-tool", "cadence-cassandra-tool"
```
Comment (carlocab) [inline Formula/cadence-workflow.rb:18]: Why can't we use the default `make` target?

File: Formula/cadence-workflow.rb [hunk 7]
```diff
@@ -0,0 +1,33 @@
+class CadenceWorkflow < Formula
+  desc "Cadence Workflow is a distributed, scalable, durable, and highly available orchestration engine to execute asynchronous long-running business logic in a scalable and resilient way"
+  homepage "https://cadenceworkflow.io/"
+  url "https://github.com/uber/cadence.git",
```
Comment (carlocab) [inline Formula/cadence-workflow.rb:4]: What is the relationship between this project and the one at https://github.com/onflow/cadence?

**Review by carlocab — DISMISSED** [review]
Thanks for your PR, @longquanzheng. Please fix the following CI errors:

```
Error: Description is too long. It should be less than 80 characters. The current length is 179.
Error: Description shouldn't start with the formula name.
Error: Line is too long. [188/118]
Error: Empty block detected.
Error: Use the `cp` Ruby method instead of `system "cp", "-r", "config", "#{etc}/cadence"`
Error: Use the `cp` Ruby method instead of `system "cp", "-r", "schema", "#{etc}/cadence"`
```

You can also view the errors locally with `brew audit --strict cadence-workflow`. (Or, better, run `brew audit --new cadence-workflow`.)

When you fix these, please squash the changes with your first commit and use the commit message
```
cadence-workflow 0.18.1 (new formula)
```

For future reference, please use the preferred commit-message style for homebrew/core. We put the name of the formula first in commit-message headings. Refer to the [documentation](https://docs.brew.sh/Formula-Cookbook#commit) for more details. This is also mentioned in the [guidelines](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md#to-add-a-new-formula-for-foo-version-234-from-url) for contributing that you read before making this PR.

File: Formula/cadence-workflow.rb [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+      tag:      "v0.18.1",
+      revision: "b6dfe37ebfec7742b1470be7129f07518e8bec98"
+  license "MIT"
+
+  bottle do
+  end
+
```
Comment (carlocab) [inline Formula/cadence-workflow.rb:11]:
```suggestion
```
You don't need this. @BrewTestBot will add the `bottle` block upon merging.

File: Formula/cadence-workflow.rb (hunk 7 above)
Comment (longquanzheng) [reply Formula/cadence-workflow.rb:4]: There is no relationship at all. 
That is a programming language for blockchain stuff, while this is a service+programming framework for orchestration/workflow. 

Unfortunately, `cadence` is a very overloaded name. There is even one more `cadence` programming https://www.cadence.com/ which they don't make homebrew yet....

Comment (carlocab) [reply Formula/cadence-workflow.rb:4]: We should probably still keep the `conflicts_with` line then.

Comment (longquanzheng) [reply Formula/cadence-workflow.rb:4]: okay.

File: Formula/cadence-workflow.rb [hunk 9]
```diff
@@ -0,0 +1,33 @@
+class CadenceWorkflow < Formula
+  desc "Cadence Workflow is a distributed, scalable, durable, and highly available orchestration engine to execute asynchronous long-running business logic in a scalable and resilient way"
+  homepage "https://cadenceworkflow.io/"
+  url "https://github.com/uber/cadence.git",
+      tag:      "v0.18.1",
+      revision: "b6dfe37ebfec7742b1470be7129f07518e8bec98"
```
Comment (carlocab) [inline Formula/cadence-workflow.rb:6]: Do we need a `git` checkout, or can we use a tarball?

File: Formula/cadence-workflow.rb (hunk 7 above)
Comment (carlocab) [reply Formula/cadence-workflow.rb:4]: Great, thanks.

When you're done making changes to this formula, and have squashed all the commits together, you'll need to push a separate commit that adds the appropriate `conflicts_with` line to the `cadence` formula as well.

File: Formula/cadence-workflow.rb (hunk 2 above)
Comment (longquanzheng) [reply Formula/cadence-workflow.rb:17]: @git-hulk can you help answer it? 
I guess it's because the binary requires libraries from GOPATH? is that correct?

File: Formula/cadence-workflow.rb (hunk 3 above)
Comment (longquanzheng) [reply Formula/cadence-workflow.rb:23]: No, but there is not such target: see https://github.com/uber/cadence/blob/master/Makefile

File: Formula/cadence-workflow.rb (hunk 6 above)
Comment (longquanzheng) [reply Formula/cadence-workflow.rb:18]: Because the default target doesn't do anything:
```
qlong@~/cadence:
(master)$make
help                 Prints a help message showing any specially-commented targets
clean                Clean binaries and build folder
copyright            update copyright headers
deps                 Check for dependency updates, for things that are directly imported
deps-all             Check for all dependency updates
fmt                  run goimports
lint                 (re)run the linter
release              Re-generate generated code and run tests
test                 Build and run all tests
qlong@~/cadence:
```

File: Formula/cadence-workflow.rb (hunk 2 above)
Comment (carlocab) [reply Formula/cadence-workflow.rb:14]: `ENV["GOPATH"]` should return `nil`, so I believe this just appends `/bin` to `PATH`, but `/bin` should already be in `PATH`.

File: Formula/cadence-workflow.rb (hunk 9 above)
Comment (longquanzheng) [reply Formula/cadence-workflow.rb:6]: Good point. Looks like we could do that. Thanks

File: Formula/cadence-workflow.rb (hunk 6 above)
Comment (longquanzheng) [reply Formula/cadence-workflow.rb:15]: Actually we could do `make bins`. Does it look better for you?

File: Formula/cadence-workflow.rb [hunk 10]
```diff
@@ -0,0 +1,29 @@
+class CadenceWorkflow < Formula
+  desc "Cadence Workflow is a distributed, scalable, durable, and highly available orchestration engine to execute asynchronous long-running business logic in a scalable and resilient way"
```
Comment (carlocab) [inline Formula/cadence-workflow.rb:2]:
```suggestion
  desc "Distributed, scalable, durable, and highly available orchestration engine"
```
This should be enough to fix the audit, I think.

File: Formula/cadence-workflow.rb (hunk 2 above)
Comment (longquanzheng) [reply Formula/cadence-workflow.rb:14]: When I search `grep GOPATH * -A 3 -B 1` in formulas, there are many different ways of doing it. 

For example, this
```
  def install
    ENV["GOPATH"] = buildpath
    ENV["GO111MODULE"] = "auto"
    ENV.prepend_create_path "PATH", buildpath/"bin"
```
does't work for us.

File: Formula/cadence-workflow.rb (hunk 10 above)
Comment (longquanzheng) [reply Formula/cadence-workflow.rb:2]: Thanks!

File: Formula/cadence-workflow.rb (hunk 2 above)
Comment (carlocab) [reply Formula/cadence-workflow.rb:14]: Yea, those existing formulae shouldn't be using `GOPATH` anymore. `GOPATH` usage is deprecated in Go (modules should be used instead), and we're no longer accepting formulae that need it in homebrew/core.

However, it looks like building using Go modules should work for this formula.

File: Formula/cadence-workflow.rb (hunk 5 above)
Comment (longquanzheng) [reply Formula/cadence-workflow.rb]: Added:
```
(cadence)$brew test cadence-workflow
==> Testing cadence-workflow
==> /usr/local/Cellar/cadence-workflow/0.18.2/bin/cadence --version
```

(btw this is a little werid that the project release version is 0.18.2 while the assertion version of `cadence` is 0.18.3
That's because we are missing the version tag in server command right now which will be fixed soon by https://github.com/uber/cadence/issues/4052
I will update this test later after that fix is released.

File: Formula/cadence-workflow.rb (hunk 7 above)
Comment (longquanzheng) [reply Formula/cadence-workflow.rb]: Added.

**Comment by longquanzheng** [comment]
Thanks @carlocab for reviewing it(on weekend, really appreciate). I have responded all your comments so far.

File: Formula/cadence-workflow.rb (hunk 2 above)
Comment (longquanzheng) [reply Formula/cadence-workflow.rb:14]: Yes, there is a script executed in Makefile to compose LDFLAG: https://github.com/uber/cadence/blob/7bd91051360214181464b87325615631a1712bb7/Makefile#L71

It's probably too complicated. Is there any concern of doing in a way of this PR?

Comment (longquanzheng) [reply Formula/cadence-workflow.rb:14]: >ENV["GOPATH"] should return nil, so I believe this just appends /bin to PATH, but /bin should already be in PATH.
Interesting, in my laptop(mac) with latest go(1.16.3), do still have that GOPATH:

```
(cadence)$go version
go version go1.16.3 darwin/amd64

(cadence)$echo $GOPATH
/Users/qlong/go

```

Comment (longquanzheng) [reply Formula/cadence-workflow.rb:14]: Actually, my laptop env doesn't need that line. Because PATH already has it.
So I think we could remove it.

Comment (longquanzheng) [reply Formula/cadence-workflow.rb]: @carlocab I removed it. LMK if you have any other suggestion. Thanks again :D

**Comment by longquanzheng** [comment]
Thanks @carlocab . I just squashed and use the commit message you suggests.

**Comment by longquanzheng** [comment]
Hi @carlocab, let me know if I still miss any of your comments. Really appreciate your review

**Comment by carlocab** [comment]
Sorry, missed your changes here. You actually need to split your commit into two: one that adds `cadence-workflow` and another that modifies `cadence`. (I actually say that `cadence` should be modified in a separate commit here: https://github.com/Homebrew/homebrew-core/pull/74494#discussion_r606700598)

As a general rule, we have only one formula file modified per commit (and only one commit per formula file modified).

File: Formula/cadence-workflow.rb [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+    bin.install "cadence-canary"
+    bin.install "cadence-sql-tool"
+    bin.install "cadence-cassandra-tool"
+
+    (etc/"cadence").install "config", "schema"
+  end
+
+  test do
+    assert_match "cadence version 0.18.3",
+      shell_output("#{bin}/cadence --version")
```
Comment (carlocab) [inline Formula/cadence-workflow.rb:26]: We still need a test more substantial than this. It has to exercise some functionality.

Please see the [docs](https://docs.brew.sh/Formula-Cookbook#add-a-test-to-the-formula) for more detail on this. Quoting from there:
>* If the formula is a library, compile and run some simple code that links against it. It could be taken from upstream's documentation / source examples.
>A good example is [`tinyxml2`](https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/tinyxml2.rb), which writes a small C++ source file into the test directory, compiles and links it against the tinyxml2 library and finally checks that the resulting program runs successfully.
>* If the formula is for a GUI program, try to find some function that runs as command-line only, like a format conversion, reading or displaying a config file, etc.
>* If the software cannot function without credentials or requires a virtual machine, docker instance, etc. to run, a test could be to try to connect with invalid credentials (or without credentials) and confirm that it fails as expected. This is prefered over mocking a dependency.
>* Homebrew comes with a number of [standard test fixtures](https://github.com/Homebrew/brew/tree/master/Library/Homebrew/test/support/fixtures), including numerous sample images, sounds, and documents in various formats. You can get the file path to a test fixture with `test_fixtures("test.svg")`.

**Comment by longquanzheng** [comment]
> Sorry, missed your changes here. You actually need to split your commit into two: one that adds `cadence-workflow` and another that modifies `cadence`. (I mentioned that `cadence` should be modified in a separate commit here: [#74494 (comment)](https://github.com/Homebrew/homebrew-core/pull/74494#discussion_r606700598))
> 
> As a general rule, we have only one formula file modified per commit (and only one commit per formula file modified).

Ahh I see. When you say a separate commit I thought you meant in the same PR(cuz in team a PR can have multiple commits). 
No problem I will do it.

File: Formula/cadence-workflow.rb (hunk 11 above)
Comment (longquanzheng) [reply Formula/cadence-workflow.rb:26]: cadence-server will need to run with databases(MySQL/Postgers or Cassandra), and cadence/cadence-canary/etc(the CLIs) will need to run with server. 

So I think I will just verify the failure case. Does it work for you? 

So without DB, the start would fail like this:
```
qlong@~/cadence:
(master)$./cadence-server start
2021/04/05 18:16:46 Loading config; env=development,zone=,configDir=./config
2021/04/05 18:16:46 Loading configFiles=[./config/base.yaml ./config/development.yaml]
2021/04/05 18:16:46 gocql: unable to dial control conn 127.0.0.1: dial tcp 127.0.0.1:9042: connect: connection refused
2021/04/05 18:16:46 cassandra schema version compatibility check failed: unable to create CQL Client: gocql: unable to create session: control: unable to connect to initial hosts: dial tcp 127.0.0.1:9042: connect: connection refused
```

Comment (carlocab) [reply Formula/cadence-workflow.rb:26]: That's fine, thanks.

Comment (longquanzheng) [reply Formula/cadence-workflow.rb]: Added. Could you take another look? Thanks

**Comment by longquanzheng** [comment]
@carlocab sorry for bothering you..What is the right way to use `assert_match`? 

If I use `assert_match /Loading config; env=development,zone=,configDir/, output`
will have error
```
Formula/cadence-workflow.rb:26:18: W: [Correctable] Ambiguous regexp literal. Parenthesize the method arguments if it's surely a regexp literal, or add a whitespace to the right of the / if it should be a division.
    assert_match /Loading config; env=development,zone=,configDir/, output
```

And if I did like it says: `assert_match (/Loading config; env=development,zone=,configDir/), output`
will get another error:
```
Formula/cadence-workflow.rb:26:18: C: [Correctable] Don't use parentheses around a literal.
    assert_match (/Loading config; env=development,zone=,configDir/), output
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

**Comment by longquanzheng** [comment]
@carlocab I finally made it passing all the checks. Could you take another look. Thanks again :D

**Comment by carlocab** [comment]
Whoops, this got buried in all the other PRs. Maybe I wasn't being very clear: we need to add a `conflicts_with` statement to `cadence` in a separate commit in this PR. I added it here for you.

**Comment by longquanzheng** [comment]
> Whoops, this got buried in all the other PRs. Maybe I wasn't being very clear: we need to add a `conflicts_with` statement to `cadence` in a separate commit in this PR. Your PR needs to modify two formula files, so it needs to have two commits.
> 
> I added it here for you.

Ahhh I finally understand what you meant.

Thank you again!

**Review by carlocab — APPROVED** [review]
Thanks, @longquanzheng.
